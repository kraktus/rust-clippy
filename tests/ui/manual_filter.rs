#![warn(clippy::manual_filter)]

fn main() {
    match Some(0) {
        None => None,
        Some(x) => {
            if x > 0 {
                None
            } else {
                Some(x)
            }
        },
    };

    match Some(1) {
        Some(x) => {
            if x > 0 {
                None
            } else {
                Some(x)
            }
        },
        None => None,
    };

    match Some(2) {
        Some(x) => {
            if x > 0 {
                None
            } else {
                Some(x)
            }
        },
        _ => None,
    };

    match Some(3) {
        Some(x) => {
            if x > 0 {
                Some(x)
            } else {
                None
            }
        },
        None => None,
    };

    let y = Some(4);
    match y {
        // Some(4)
        None => None,
        Some(x) => {
            if x > 0 {
                None
            } else {
                Some(x)
            }
        },
    };

    match Some(5) {
        Some(x) => {
            if x > 0 {
                Some(x)
            } else {
                None
            }
        },
        _ => None,
    };

    
    match Some(6) { // Not supported for now, but it could lint and remove `ref`, because this pattern only work if `x` implements the `Copy` trait.
        Some(ref x) => {
            if x > &0 {
                Some(x)
            } else {
                None
            }
        },
        _ => None,
    };

    let external_cond = true;
    match Some(String::new()) {
        Some(x) => {
            if external_cond {
                Some(x)
            } else {
                None
            }
        },
        _ => None,
    };

    if let Some(x) = Some(7) {
        if external_cond {
                Some(x)
            } else {
                None
            }
    } else {
        None
    };

    match &Some(8) {
        &Some(x) => if x != 0 {
                Some(x)
            } else {
                None
            },
        _ => None,
    };
}
