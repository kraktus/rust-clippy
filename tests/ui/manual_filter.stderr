error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:4:5
   |
LL | /     match Some(0) {
LL | |         None => None,
LL | |         Some(x) => {
LL | |             if x > 0 {
...  |
LL | |         },
LL | |     };
   | |_____^ help: try: `Some(0).filter(|&x| x <= 0)`
   |
   = note: `-D clippy::manual-filter` implied by `-D warnings`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:15:5
   |
LL | /     match Some(1) {
LL | |         Some(x) => {
LL | |             if x > 0 {
LL | |                 None
...  |
LL | |         None => None,
LL | |     };
   | |_____^ help: try: `Some(1).filter(|&x| x <= 0)`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:26:5
   |
LL | /     match Some(2) {
LL | |         Some(x) => {
LL | |             if x > 0 {
LL | |                 None
...  |
LL | |         _ => None,
LL | |     };
   | |_____^ help: try: `Some(2).filter(|&x| x <= 0)`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:37:5
   |
LL | /     match Some(3) {
LL | |         Some(x) => {
LL | |             if x > 0 {
LL | |                 Some(x)
...  |
LL | |         None => None,
LL | |     };
   | |_____^ help: try: `Some(3).filter(|&x| x > 0)`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:49:5
   |
LL | /     match y {
LL | |         // Some(4)
LL | |         None => None,
LL | |         Some(x) => {
...  |
LL | |         },
LL | |     };
   | |_____^ help: try: `y.filter(|&x| x <= 0)`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:61:5
   |
LL | /     match Some(5) {
LL | |         Some(x) => {
LL | |             if x > 0 {
LL | |                 Some(x)
...  |
LL | |         _ => None,
LL | |     };
   | |_____^ help: try: `Some(5).filter(|&x| x > 0)`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:85:5
   |
LL | /     match Some(String::new()) {
LL | |         Some(x) => {
LL | |             if external_cond {
LL | |                 Some(x)
...  |
LL | |         _ => None,
LL | |     };
   | |_____^ help: try: `Some(String::new()).filter(|&x| external_cond)`

error: manual implementation of `Option::filter`
  --> $DIR/manual_filter.rs:96:5
   |
LL | /     if let Some(x) = Some(7) {
LL | |         if external_cond {
LL | |                 Some(x)
LL | |             } else {
...  |
LL | |         None
LL | |     };
   | |_____^ help: try: `Some(7).filter(|&x| external_cond)`

error: aborting due to 8 previous errors

